### ISSUE TITLE
Keep Spaces/Comments/Layout

### ISSUE THREAD

**Issue by jogibear9988** [csstree/csstree#237]
We now use csstree in our web component designer (https://node-projects.github.io/web-component-designer-demo/index.html). It works very well (for the first draft), but now we have the problem that all the formating of the css is lost.

Is there a way to keep all the spaces/linebrakes and comments?

**Comment by bartveneman** [comment]
Seems like a duplicate of #78 perhaps?

**Comment by scripthunter7** [comment]
Yes, duplicate. I also opened a similar issue, it was also a duplicate: https://github.com/csstree/csstree/issues/212

It would be nice to come up with some kind of algorithm / data structure for this, since keeping the formatting would be useful in many cases.

**Comment by bartveneman** [comment]
I posted a workaround in that one: https://github.com/csstree/csstree/issues/78#issuecomment-1368057828

**Comment by scripthunter7** [comment]
I saw it, it's an interesting approach :) I also tried to solve it based on the locations. I use this library for parsing the selectors of filter rules in adblock filter lists, so line breaks are irrelevant to me (since an adblock rule is fixedly 1 line), only the spaces count, and that could be calculated from the two also from the distance between nodes (although this approach does not handle the space-tab difference).

The disadvantage of our approaches is that they use "side effects" and do not consistently store the information in the nodes / AST (like PostCSS, for example).

In my opinion, this does not even mean a breaking change, since the raws field is only additional information, the current AST remains compatible.

**Comment by jogibear9988** [comment]
I now switched to https://github.com/adobe/css-tools in my designer. It also looses some whitespace & formating, but at least it keeps the comments
